cpp/include/raft/sparse/matrix/detail/preprocessing.cuh

```diff
@@ -0,0 +1,463 @@
+ * @param data: Input COO values array.
+ * @param itr_vals: Input array used to calculate counts.
+ * @param keys_out: Output array with one entry for each key. (same size as counts_out)
+ * @param counts_out: Output array with cumulative sum for each key. (same size as keys_out)
+ */
+template <typename T1, typename T2, typename IdxT>
+void get_uniques_counts(raft::resources& handle,
+                        raft::device_vector_view<T1, IdxT> sort_vector,
+                        raft::device_vector_view<T1, IdxT> secondary_vector,
+                        raft::device_vector_view<T2, IdxT> data,
+                        raft::device_vector_view<T2, IdxT> itr_vals,
+                        raft::device_vector_view<T1, IdxT> keys_out,
+                        raft::device_vector_view<T2, IdxT> counts_out)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  raft::sparse::op::coo_sort(sort_vector.size(),
+                             secondary_vector.size(),
+                             data.size(),
+                             sort_vector.data_handle(),
+                             secondary_vector.data_handle(),
+                             data.data_handle(),
+                             stream);
+
+  thrust::reduce_by_key(raft::resource::get_thrust_policy(handle),
+                        sort_vector.data_handle(),
+                        sort_vector.data_handle() + sort_vector.size(),
+                        itr_vals.data_handle(),
+                        keys_out.data_handle(),
+                        counts_out.data_handle());
+}
+
+/**
+ * @brief Compute cumulative sum for each unique value in the origin array
+ * @param handle: raft resource handle
+ * @param origin: Input array that has values to use for computation
+ * @param keys: Output array that has keys, should be the size of unique
+ * @param counts: Output array that contains the computed counts
+ * @param results: Output array that scatters the counts to origin value positions. Same size as
+ * origin array.
+ */
+template <typename T1, typename T2, typename IdxT>
+void create_mapped_vector(raft::resources& handle,
+                          const raft::device_vector_view<T1, IdxT> origin,
+                          const raft::device_vector_view<T1, IdxT> keys,
+                          const raft::device_vector_view<T2, IdxT> counts,
+                          raft::device_vector_view<T2, IdxT> result)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  auto host_keys      = raft::make_host_vector<T1, IdxT>(handle, keys.size());
+
+  raft::copy(host_keys.data_handle(), keys.data_handle(), keys.size(), stream);
+  raft::linalg::map(handle, result, raft::cast_op<T2>{}, raft::make_const_mdspan(origin));
+  int new_key_size = host_keys(host_keys.size() - 1) + 1;
+  auto origin_map  = raft::make_device_vector<T2, IdxT>(handle, new_key_size);
+
+  thrust::scatter(raft::resource::get_thrust_policy(handle),
```

Comment:
Probably want to avoid using thrust directly as much as possible and use raft primitives. I believe we have a scatter primitive. Will [this](https://github.com/rapidsai/raft/blob/branch-24.10/cpp/include/raft/matrix/scatter.cuh) work?

Reply:
Unfortunately I cannot figure out how to use the raft::scatter function for this because of the way the kernelscatter is built. It expects an op and it basically maps the index of the input to the index of the output. I need something slightly different, where I need the value of the input to be placed in the respective index. (i.e. [2 ,3 ,4], [1, 2, 1] => [0, 0, 1, 2 ,1]) the way raft::scatter works (i.e. [2 ,3 ,4], [1, 2, 1] => [1,2,1, 0, 0] ) and because of the mapping here https://github.com/rapidsai/raft/blob/branch-24.10/cpp/include/raft/util/detail/scatter.cuh#L38 it ties whatever comes out of an index in the input data to that index in the output (at least that's how I understand it).

Reply:
no longer an issue this scatter was replaced by a map call using raft api